Sources/KSCrashRecording/KSCrashAppMemory.mm

```diff
@@ -0,0 +1,362 @@
+#import "KSCrashAppMemory.h"
+
+#import <mach/mach.h>
+#import <mach/task.h>
+#import <atomic>
+
+#if TARGET_OS_IOS
+#import <UIKit/UIKit.h>
+#endif
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ The memory tracker takes care of centralizing the knowledge around memory.
+ It does the following:
+ 
+ 1- Wraps memory pressure. This is more useful than `didReceiveMemoryWarning`
+ as it vends different levels of pressure caused by the app as well as the rest of the OS.
+ 
+ 2- Vends a memory level. This is pretty novel. It vends levels of where the app is wihtin
+ the memory limit.
+ 
+ Some useful info.
+ 
+ Memory Pressure is mostly useful when the app is in the background.
+ It helps understand how much `pressure` is on the app due to external concerns. Using
+ this data, we can make informed decisions around the reasons the app might have been
+ terminated.
+ 
+ Memory Level is useful in the foreground as well as background. It indicates where the app is
+ within its memory limit. That limit being calculated by the addition of `remaining` and
+ `footprint`. Using this data, we can also make informaed decisions around foreground and background
+ memory terminations, aka. OOMs.
+ 
+ See: https://github.com/naftaly/Footprint
+ */
+
+@interface KSCrashAppMemoryTracker () {
+    dispatch_queue_t _heartbeatQueue;
+    dispatch_source_t _pressureSource;
+    dispatch_source_t _limitSource;
+    std::atomic<uint64_t> _footprint;
```

Comment:
I expect that OSAtomic would do just as good. I chose std::atomic for its simplicity. I rarely use atomic properties since they are usually under spin locks (I think) which can lead to problems such as priority inversion.I’m happy to follow suite with how the system prefers it.

Reply:
As I was converting this to atomic properties it dawned on me that they don't have "exchange" methods which can often make the code a lot more complicated.

Reply:
Oh, I see. Maybe `OSAtomicCompareAndSwap64` would be a good replacement. I get that it won't have the same type safety and simplicity, but it's a good trade-off to avoid using C++.

Reply:
I've moved away from atomics in favor of unfair locks. The multiple atomics plus the need for swapping values make locks a better tool to use in this case.